ribs/archives/_dns_archive.py

```diff
@@ -0,0 +1,587 @@
+            cur = {
+                name: np.empty((0, *desc[0]), dtype=desc[1])
+                for name, desc in self._store.field_desc.items()
+            }  # type: ignore[index]
+
+        # Combine.
+        combined = {}
+        for name in self._store.field_list:
+            combined[name] = (
+                np.concatenate((cur[name], data[name]), axis=0)
+                if name in data
+                else cur[name]
+            )
+
+        # Compute DNS over union.
+        measures = combined["measures"]
+        objs = combined["objective"]
+
+        # Pairwise distances.
+        if measures.shape[0] == 0:
+            dns_scores = np.array([], dtype=self.dtypes["measures"])  # pragma: no cover
+        else:
+            dist = cdist(measures, measures)
+            # Exclude self distances.
+            np.fill_diagonal(dist, np.inf)
+            # For each i, consider only fitter neighbors (obj[j] >= obj[i]).
+            fitter_mask = objs[None, :] >= objs[:, None]
+            # Replace non-fitter distances with +inf.
+            dist_masked = np.where(fitter_mask, dist, np.inf)
+            # Compute mean of k smallest along each row (ignoring +inf when <k exist).
+            dns_scores = np.empty(measures.shape[0], dtype=self.dtypes["measures"])
+            for i in range(measures.shape[0]):
+                row = dist_masked[i]
+                valid = np.isfinite(row)
+                if not np.any(valid):
+                    dns_scores[i] = np.inf
+                    continue
+                k = min(self._k_neighbors, np.sum(valid))
+                part = np.partition(row[valid], k - 1)[:k]
+                dns_scores[i] = np.mean(part)
+
+        # Select survivors: top `capacity` by DNS (descending).
+        cap = self.capacity
+        n_total = dns_scores.shape[0]
+        if n_total <= cap:
+            survivor_indices = np.arange(n_total)
+        else:
+            # Take largest `cap` values.
+            survivor_indices = np.argpartition(dns_scores, -cap)[-cap:]
+            survivor_indices = survivor_indices[
+                np.argsort(dns_scores[survivor_indices])
+            ]
+
+        # Build add_info for batch entries.
+        batch_size = len(data["measures"])
+        add_info = {
+            "status": np.zeros(batch_size, dtype=np.int32),
+            "dns": np.empty(batch_size, dtype=self.dtypes["measures"]),
+        }
+        batch_start = cur_size
+        batch_indices_in_union = np.arange(batch_start, batch_start + batch_size)
+        batch_survivors = np.isin(batch_indices_in_union, survivor_indices)
+        add_info["status"][batch_survivors] = 2
+        add_info["dns"] = dns_scores[batch_indices_in_union]
+
+        # Rebuild population as survivors (compact to [0..n-1]).
+        survivors = {
+            name: combined[name][survivor_indices] for name in self._store.field_list
+        }
+        self._store.clear()
+        if survivors["measures"].shape[0] > 0:
```

Comment:
Repeatedly clearing and adding the store seems like it is prone to performance issues, but I can't really think of better ways at the moment. Maybe try running DNS on larger problems and we can circle back if this turns out to be a bottleneck.